ratelimits/limiter.go

```diff
@@ -238,90 +164,212 @@ func (l *Limiter) Spend(ctx context.Context, txn Transaction) (*Decision, error)
 	return d, nil
 }
 
+func prepareBatch(txns []Transaction) ([]Transaction, []string, error) {
+	var bucketKeys []string
+	var transactions []Transaction
+	for _, txn := range txns {
+		if txn.allowOnly() {
+			// Ignore allow-only transactions.
+			continue
+		}
+		if slices.Contains(bucketKeys, txn.bucketKey) {
+			return nil, nil, fmt.Errorf("found duplicate bucket %q in batch", txn.bucketKey)
+		}
+		bucketKeys = append(bucketKeys, txn.bucketKey)
+		transactions = append(transactions, txn)
+	}
+	return transactions, bucketKeys, nil
+}
+
+type batchDecision struct {
+	*Decision
+}
+
+func newBatchDecision() *batchDecision {
+	return &batchDecision{
+		Decision: &Decision{
+			Allowed:   true,
+			Remaining: math.MaxInt64,
+		},
+	}
+}
+
+func (d *batchDecision) merge(in *Decision) {
+	d.Allowed = d.Allowed && in.Allowed
+	d.Remaining = min(d.Remaining, in.Remaining)
+	d.RetryIn = max(d.RetryIn, in.RetryIn)
+	d.ResetIn = max(d.ResetIn, in.ResetIn)
+	if in.newTAT.After(d.newTAT) {
+		d.newTAT = in.newTAT
+	}
+}
+
+// BatchSpend attempts to deduct the costs from the provided buckets'
+// capacities. If applicable, new bucket states are persisted to the underlying
+// datastore before returning. Non-existent buckets will be initialized WITH the
+// cost factored into the initial state. The following rules are applied to
+// merge the Decisions for each Transaction into a single batch Decision:
+//   - Allowed is true if all Transactions where check is true were allowed,
+//   - RetryIn and ResetIn are the largest values of each across all Decisions,
+//   - Remaining is the smallest value of each across all Decisions, and
+//   - Decisions resulting from spend-only Transactions are never merged.
+func (l *Limiter) BatchSpend(ctx context.Context, txns []Transaction) (*Decision, error) {
+	batch, bucketKeys, err := prepareBatch(txns)
+	if err != nil {
+		return nil, err
+	}
+	if len(batch) == 0 {
+		// All Transactions were allow-only.
+		return allowedDecision, nil
+	}
+
+	// Remove cancellation from the request context so that transactions are not
+	// interrupted by a client disconnect.
+	ctx = context.WithoutCancel(ctx)
+	tats, err := l.source.BatchGet(ctx, bucketKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	start := l.clk.Now()
+	batchDecision := newBatchDecision()
+	newTATs := make(map[string]time.Time)
+
+	for _, txn := range batch {
+		tat, exists := tats[txn.bucketKey]
+		if !exists {
+			// First request from this client.
+			tat = l.clk.Now()
+		}
+
+		d := maybeSpend(l.clk, txn.limit, tat, txn.cost)
+
+		if txn.limit.isOverride {
+			utilization := float64(txn.limit.Burst-d.Remaining) / float64(txn.limit.Burst)
+			l.overrideUsageGauge.WithLabelValues(txn.limit.name.String(), txn.bucketKey).Set(utilization)
+		}
+
+		if d.Allowed && (tat != d.newTAT) && txn.spend {
+			// New bucket state should be persisted.
+			newTATs[txn.bucketKey] = d.newTAT
+		}
+
+		if txn.spendOnly() {
+			d = allowedDecision
+		}
+		batchDecision.merge(d)
```

Comment:
This could be simplified somewhat:

```suggestion
		if !txn.spendOnly() {
		  batchDecision.merge(d)
		}
```